Add zippering support

etcwilde opened the pull request:
Setting the `SWIFT_ENABLE_MACCATALYST` flag will build the compatibility libraries with zippering enabled.

Note that AppleClang and LLVM Clang use different flags to write zippered files. The Swift stdlib build doesn't know what clang we're using, so we can't ask it for the compiler ID to determine which whether we're using AppleClang or LLVM clang, hence the nasty `execute_process` trick to figure out whether the compiler that's actually compiling the stdlib knows about the flag or not.

This should fix #61418.

etcwilde committed:
Add zippering support

Upstream build system zippering support.

Setting the `SWIFT_ENABLE_MACCATALYST` flag will build the compatibility
libraries with zippering enabled.

Note that AppleClang and LLVM Clang use different flags to write
zippered files. The Swift stdlib build doesn't know what clang we're
using, so we can't ask it for the compiler ID to determine which whether
we're using AppleClang or LLVM clang, hence the nasty `execute_process`
trick to figure out whether the compiler that's actually compiling the
stdlib knows about the flag or not.

etcwilde commented:
@compnerd I'm pretty sure that this shouldn't break Windows since the runtimes should only be built with the special clang, which will know about the catalyst bundling and because we can't build for catalyst on Windows anyway. But I'm not 100% certain. I'd appreciate your input.

etcwilde commented:
@swift-ci please test

compnerd commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -82,7 +82,28 @@ function(_add_target_variant_c_compile_link_flags)
+      # Check if the C compiler supports `-target-variant` flag
+      # TODO (etcwilde): This is a massive hack to deal with the fact that we
+      # are lying to cmake about what compiler is being used. Normally we could
+      # use `check_cxx_compiler_flag` here. Unfortunately, that uses a different
```
```suggestion
      # use `check_c_compiler_flag` here. Unfortunately, that uses a different
```

etcwilde replied:
I think it's `check_compiler_flag(<lang> ...)` now anyway. Will fix comment.

compnerd commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -82,7 +82,28 @@ function(_add_target_variant_c_compile_link_flags)
+        RESULT_VARIABLE
+          SUPPORTS_TARGET_VARIANT)
+
+      if(SUPPORTS_TARGET_VARIANT EQUAL 0)
```
```suggestion
      if(NOT SUPPORTS_TARGET_VARIANT)
```

compnerd commented on stdlib/cmake/modules/AddSwiftStdlib.cmake:
```diff
@@ -82,7 +82,28 @@ function(_add_target_variant_c_compile_link_flags)
+      # TODO (etcwilde): This is a massive hack to deal with the fact that we
+      # are lying to cmake about what compiler is being used. Normally we could
+      # use `check_cxx_compiler_flag` here. Unfortunately, that uses a different
+      # compiler since we swap out the C compiler part way through the build.
```
I really wish that we would stop doing this ...

etcwilde replied:
I do too. I do too.

compnerd reviewed (approved):
This should be safe for Windows, but this is still pretty terrible.  The compiler swap occurs at stdlib build time right?

etcwilde commented:
> The compiler swap occurs at stdlib build time right?

It happens in a couple of places. We swap the compilers for the stdlib, but it's also done in the bootstrap IIRC.

etcwilde committed:
Build with catalyst support in nightly osx package

Adding maccatalyst support to the osx nightly package.

etcwilde commented:
This is looking pretty good now. My local test with catalyst enabled is linking correctly but I'm still missing the catalyst swift modules. I'll keep digging.

etcwilde committed:
Zipper all the things!

This patch goes through and adds zippering and the swift module
dependencies to a bunch of pieces of the swift runtimes. Here's to
hoping I hit everything that needed to be hit. :D

With this patch, I'm seeing the appropriate modules under
lib/swift/maccatalyst, so things seem to be working right.

etcwilde commented:
@swift-ci please test

etcwilde commented:
@swift-ci please macOS toolchain build

etcwilde commented:
@swift-ci Please Build Toolchain macOS Platform

etcwilde commented:
@swift-ci please test

etcwilde commented:
Linux failure:
```
warning: 'swift-llbuild': skipping cache due to an error: Failed to clone repository https://github.com/apple/swift-llbuild.git:
    Cloning into bare repository '/home/build-user/build/buildbot_linux/.cache/org.swift.swiftpm/repositories/swift-llbuild-0f2964f9'...
    fatal: unable to access 'https://github.com/apple/swift-llbuild.git/': Failed to connect to github.com port 443: Connection timed out
```

Unrelated CI failure.

etcwilde commented:
@swift-ci please test Linux
